chainbench/util/monitor.py

```diff
@@ -0,0 +1,107 @@
+    Therefore, we use max function to ensure the lag calculated is minimum 0 and never
+    negative.
+    """
+    return max(int((current_timestamp - block_timestamp).total_seconds()), 0)
+
+
+def parse_timespan(time_str):
+    """
+    Parse a string representing a time span and return the number of seconds.
+    Valid formats are: 20, 20s, 3m, 2h, 1h20m, 3h30m10s, etc.
+    """
+    if not time_str:
+        raise ValueError("Invalid time span format")
+
+    if re.match(r"^\d+$", time_str):
+        # if an int is specified we assume they want seconds
+        return int(time_str)
+
+    timespan_regex = re.compile(
+        r"((?P<hours>\d+?)h)?((?P<minutes>\d+?)m)?((?P<seconds>\d+?)s)?"
+    )
+    parts = timespan_regex.match(time_str)
+    if not parts:
+        raise ValueError(
+            "Invalid time span format. "
+            "Valid formats: 20, 20s, 3m, 2h, 1h20m, 3h30m10s, etc."
+        )
+    parts = parts.groupdict()
+    time_params = {name: int(value) for name, value in parts.items() if value}
+    if not time_params:
+        raise ValueError(
+            "Invalid time span format. "
+            "Valid formats: 20, 20s, 3m, 2h, 1h20m, 3h30m10s, etc."
+        )
+    return int(timedelta(**time_params).total_seconds())
+
+
+def head_lag_monitor(endpoint, result_path, duration):
+    data = {
+        "id": 1,
+        "jsonrpc": "2.0",
+        "method": "eth_getBlockByNumber",
+        "params": ["latest", False],
+    }
+
+    csv_writer_kwargs = {
+        "file": f"{result_path}/head_lag.csv",
+        "mode": "a",
+        "encoding": "utf-8-sig",
+        "newline": "",
+    }
+    logger.info("Creating head_lag.csv")
+    with open(**csv_writer_kwargs) as csv_file:
+        csv_writer = csv.writer(csv_file)
+        csv_writer.writerow(["timestamp", "lag (s)", "block number"])
+        logger.info("head_lag.csv created")
+    end_time = datetime.now() + timedelta(seconds=parse_timespan(duration))
+
+    http = httpx.Client()
+
+    logger.info("Start monitoring head lag")
+    while datetime.now() < end_time:
+        current_timestamp = datetime.now()
+        response = http.post(
+            endpoint,
+            json=data,
+        )
+        try:
+            block_timestamp = datetime.fromtimestamp(
+                int(response.json()["result"]["timestamp"], 0)
+            )
+            block_number = int(response.json()["result"]["number"], 0)
+            with open(**csv_writer_kwargs) as csv_file:
+                csv_writer = csv.writer(csv_file)
+                logger.info("Write row to head_lag.csv")
+                csv_writer.writerow(
+                    [
+                        current_timestamp,
+                        f"{calculate_lag(current_timestamp, block_timestamp)}",  # noqa E501
+                        block_number,
+                    ]
+                )
+                logger.info("Written 1 row to head_lag.csv")
```

Comment:
This could be a helper function as same code used twice

Reply:
refactored the code :)